File: lttng/org.lttng.scope.lttng.kernel.core/src/org/lttng/scope/lttng/kernel/core/analysis/os/handlers/internal/ProcessForkHandler.java
/*******************************************************************************
 * Copyright (c) 2015 Ericsson
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    Matthew Khouzam - Initial API and implementation
 *******************************************************************************/

package org.lttng.scope.lttng.kernel.core.analysis.os.handlers.internal;

import static java.util.Objects.requireNonNull;

import org.lttng.scope.lttng.kernel.core.analysis.os.Attributes;
import org.lttng.scope.lttng.kernel.core.analysis.os.StateValues;
import org.lttng.scope.lttng.kernel.core.trace.layout.ILttngKernelEventLayout;

import com.efficios.jabberwocky.trace.event.FieldValue.IntegerValue;
import com.efficios.jabberwocky.trace.event.FieldValue.StringValue;
import com.efficios.jabberwocky.trace.event.ITraceEvent;

import ca.polymtl.dorsal.libdelorean.ITmfStateSystemBuilder;
import ca.polymtl.dorsal.libdelorean.exceptions.AttributeNotFoundException;
import ca.polymtl.dorsal.libdelorean.statevalue.ITmfStateValue;
import ca.polymtl.dorsal.libdelorean.statevalue.TmfStateValue;

/**
 * Fork Handler
 */
public class ProcessForkHandler extends KernelEventHandler {

    /**
     * Constructor
     *
     * @param layout
     *            event layout
     */
    public ProcessForkHandler(ILttngKernelEventLayout layout) {
        super(layout);
    }

    @Override
    public void handleEvent(ITmfStateSystemBuilder ss, ITraceEvent event) throws AttributeNotFoundException {
        int cpu = event.getCpu();
        String childProcessName = requireNonNull(event.getField(getLayout().fieldChildComm(), StringValue.class)).getValue();
        int parentTid = Long.valueOf(requireNonNull(event.getField(getLayout().fieldParentTid(), IntegerValue.class)).getValue()).intValue();
        int childTid = Long.valueOf(requireNonNull(event.getField(getLayout().fieldChildTid(), IntegerValue.class)).getValue()).intValue();

        String parentThreadAttributeName = Attributes.buildThreadAttributeName(parentTid, cpu);
        if (parentThreadAttributeName == null) {
            return;
        }

        String childThreadAttributeName = Attributes.buildThreadAttributeName(childTid, cpu);
        if (childThreadAttributeName == null) {
            return;
        }

        final int threadsNode = KernelEventHandlerUtils.getNodeThreads(ss);
        Integer parentTidNode = ss.getQuarkRelativeAndAdd(threadsNode, parentThreadAttributeName);
        Integer childTidNode = ss.getQuarkRelativeAndAdd(threadsNode, childThreadAttributeName);


        /* Assign the PPID to the new process */
        int quark = ss.getQuarkRelativeAndAdd(childTidNode, Attributes.PPID);
        ITmfStateValue value = TmfStateValue.newValueInt(parentTid);
        long timestamp = event.getTimestamp();
        ss.modifyAttribute(timestamp, value, quark);

        /* Set the new process' exec_name */
        quark = ss.getQuarkRelativeAndAdd(childTidNode, Attributes.EXEC_NAME);
        value = TmfStateValue.newValueString(childProcessName);
        ss.modifyAttribute(timestamp, value, quark);

        /*
         * Set the new process' status, it is initially in a blocked state. A
         * subsequent sched_wakeup_new will schedule it.
         */
        value = StateValues.PROCESS_STATUS_WAIT_BLOCKED_VALUE;
        ss.modifyAttribute(timestamp, value, childTidNode);

        /* Set the process' syscall name, to be the same as the parent's */
        quark = ss.getQuarkRelativeAndAdd(parentTidNode, Attributes.SYSTEM_CALL);
        value = ss.queryOngoingState(quark);
        if (!value.isNull()) {
            quark = ss.getQuarkRelativeAndAdd(childTidNode, Attributes.SYSTEM_CALL);
            ss.modifyAttribute(timestamp, value, quark);
        }

    }
}

File: jabberwocky-lttng/src/main/kotlin/com/efficios/jabberwocky/lttng/kernel/analysis/os/handlers/ProcessForkHandler.kt
/*
 * Copyright (C) 2018 EfficiOS Inc., Alexandre Montplaisir <alexmonthy@efficios.com>
 * Copyright (C) 2015 Ericsson
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

package com.efficios.jabberwocky.lttng.kernel.analysis.os.handlers

import ca.polymtl.dorsal.libdelorean.IStateSystemWriter
import ca.polymtl.dorsal.libdelorean.statevalue.StateValue
import com.efficios.jabberwocky.lttng.kernel.analysis.os.Attributes
import com.efficios.jabberwocky.lttng.kernel.analysis.os.StateValues
import com.efficios.jabberwocky.lttng.kernel.trace.layout.LttngKernelEventLayout
import com.efficios.jabberwocky.trace.event.FieldValue.IntegerValue
import com.efficios.jabberwocky.trace.event.FieldValue.StringValue
import com.efficios.jabberwocky.trace.event.TraceEvent

class ProcessForkHandler(layout: LttngKernelEventLayout) : KernelEventHandler(layout) {

    override fun handleEvent(ss: IStateSystemWriter, event: TraceEvent) {
        val cpu = event.cpu
        val timestamp = event.timestamp

        val childProcessName = (event.fields[layout.fieldChildComm] as StringValue).value
        val parentTid = (event.fields[layout.fieldParentTid] as IntegerValue).value.toInt()
        val childTid = (event.fields[layout.fieldChildTid] as IntegerValue).value.toInt()

        val parentThreadAttributeName = Attributes.buildThreadAttributeName(parentTid, cpu) ?: return
        val childThreadAttributeName = Attributes.buildThreadAttributeName(childTid, cpu) ?: return

        val threadsNode = ss.getNodeThreads()
        val parentTidNode = ss.getQuarkRelativeAndAdd(threadsNode, parentThreadAttributeName)
        val childTidNode = ss.getQuarkRelativeAndAdd(threadsNode, childThreadAttributeName)


        /* Assign the PPID to the new process */
        ss.modifyAttribute(timestamp,
                StateValue.newValueInt(parentTid),
                ss.getQuarkRelativeAndAdd(childTidNode, Attributes.PPID))

        /* Set the new process' exec_name */
        ss.modifyAttribute(timestamp,
                StateValue.newValueString(childProcessName),
                ss.getQuarkRelativeAndAdd(childTidNode, Attributes.EXEC_NAME))

        /*
         * Set the new process' status, it is initially in a blocked state. A
         * subsequent sched_wakeup_new will schedule it.
         */
        ss.modifyAttribute(timestamp,
                StateValues.PROCESS_STATUS_WAIT_BLOCKED_VALUE,
                childTidNode)

        /* Set the process' syscall name, to be the same as the parent's */
        ss.getQuarkRelativeAndAdd(parentTidNode, Attributes.SYSTEM_CALL)
                .let { quark -> ss.queryOngoingState(quark) }
                .let { sv ->
                    if (!sv.isNull) {
                        ss.modifyAttribute(timestamp,
                                sv,
                                ss.getQuarkRelativeAndAdd(childTidNode, Attributes.SYSTEM_CALL))
                    }
                }

    }
}
